### PR TITLE
Cache and singleflight requests to kubelet

### LINKED ISSUE #5111: Consider caching the Kubelet response
The Kubernetes Workload Attestor queries the Kubelet for pod information on each attestation attempt. Kubelet responses can be somewhat large (e.g. 1MiB). When many processes are concurrently under attestation, the large response can cause a spike in SPIRE agent memory usage.

This issue is to explore if and how we cache the kubelet response to share across attestation attempts.

Some considerations:
1. How long is this cached for?
2. What events would actively invalidate the cache ahead of expiration? (e.g. expected pod/container not found in response)
3. Each attestation invocation is looking for a different pod/container. Do we cache just the kubelet response and then use targeted parsing of the response to cherry-pick the specific pod/container (what we do today)? Or do we parse the whole response and then share the parsed version (maybe reduces load due to re-parsing)?

Any decisions about the considerations above should ideally be backed by benchmarking to understand the tradeoffs.

### PR DESCRIPTION
**Pull request by knisbet**
Add microcaching and merging of parallel requests to kubelet in the k8s workload attestor.

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**

Reduces memory usage of the agent plugin for k8s workload attestation. #5111

**Description of change**

- Add a micro cache of half the retry interval to reduce the requests to kubelet and parsing of JSON response.
- Use x/sync/singleflight to merge parallel kubelet requests into a single request
- Scope creep: Add a context with basic 10s timeout on HTTP requests to Kubelet

**Which issue this PR fixes**

Fixes #5111

### CODE REVIEW

**Comment by knisbet** [comment]
I hope no one minds, I wanted to play around with Spire on the weekend and I saw the issue about bursts in memory usage that didn't seem to have any activity since May. Thought I'd hack on the issue a bit and see what turned out.

Testing was done by scaling the deployment in https://github.com/spiffe/spire-tutorials/blob/main/k8s/quickstart/client-deployment.yaml to 100 replicas. And a delete of all pods in the namespace a couple of times.

On my homelab:
Peak Memory usage went from 513MiB to 57 MiB
Number of requests to kubelet went from 3028 to 103

Notes:
- I suspect this may also impact #5109 as the json parsing library isn't as heavily used with this change. 
- By caching for half the retry I don't see a reason that more complicated expiry behaviour would be needed as suggested in #5111, as subsequent retries should never use the same cached request. But please check that incase I missed a reason.
- Last, I couldn't find a context or timeout on the kubelet http client. I might've missed it somewhere, so I threw one in with a fairly generous 10 second timeout to ensure a timeout was getting applied.

File: pkg/agent/plugin/workloadattestor/k8s/k8s.go
Comment (azdagron) [inline pkg/agent/plugin/workloadattestor/k8s/k8s.go:436]: In practice, kicking off this goroutine without any soft of lifetime management should be ok given its current responsibilities, but in general we push for goroutine hygiene as much as possible. If we can, it would be good to make sure that this goroutine does not outlive the plugin, e.g. when the plugin is being unloaded.

One way to accomplish this is to:
1. Implement io.Closer on *Plugin. The plugin framework will invoke Close() when the plugin is unloaded
2. Add a wait group to the plugin that tracks the lifetime of this goroutine that close waits for

It might be overkill, considering this goroutine will only live at most ~250ms at the time of close, but bonus points, if you wanted to add a context that could be cancelled inside of Close to immediately bring it down.

**Comment by azdagron** [comment]
Thanks very much for this @knisbet! There are several people who will be very excited for this change. I dropped one comment on goroutine lifetimes but other than that this is looking great.

File: pkg/agent/plugin/workloadattestor/k8s/k8s.go
Comment (knisbet) [reply pkg/agent/plugin/workloadattestor/k8s/k8s.go:436]: Agreed, I should be able to post a fix this afternoon. I didn't bother for the same reasons you indicated, shouldn't take long to add in and test.

Comment (knisbet) [reply pkg/agent/plugin/workloadattestor/k8s/k8s.go:452]: I added the io.Closer and a context. I added the Context to the struct, but understand that can be debatable in different projects, happy to swap it out if there is a preference not to use a context within a struct in this project.

Also, I didn't link the Close function to wait for any in progress Attest calls to complete. I didn't see the behaviour in any other plugins, and I didn't deeply investigate whether the service is cancelling and waiting for any other in-progress Attest calls before unloading plugins. If an Attest call is able to still be in progress there is a small window where the cache will be saved and then immediately released. I'm happy to address if you'd like.

**Comment by knisbet** [comment]
@azdagron I think I addressed your concern. Also, I don't recall seeing anywhere in the committer docs if there is a preference to Squash my changes into a single Commit. Let me know.

**Review by azdagron — APPROVED** [review]
Looks great! Thanks again, @knisbet.

**Comment by azdagron** [comment]
As far as squashing, the maintainers typically do that when we merge.

SPIRE requires DCO sign-off on all commits. Would you mind amending your commits and adding that?

**Comment by knisbet** [comment]
> SPIRE requires DCO sign-off on all commits. Would you mind amending your commits and adding that?

Oops, sorry my bad. Should be fixed now.

**Pull request merged**
